tools/talis/google_cloud.go

```diff
@@ -0,0 +1,757 @@
+	return err
+}
+
+func (c *GCClient) List(ctx context.Context) error {
+	opts, err := gcClientOptions(c.cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create client options: %w", err)
+	}
+
+	client, err := compute.NewInstancesRESTClient(ctx, opts...)
+	if err != nil {
+		return fmt.Errorf("failed to create compute client: %w", err)
+	}
+	defer client.Close()
+
+	cnt := 0
+	for _, region := range GCRegions {
+		zones := GCZones[region]
+		for _, zone := range zones {
+			req := &computepb.ListInstancesRequest{
+				Project: c.project,
+				Zone:    zone,
+			}
+			it := client.List(ctx, req)
+			for {
+				instance, err := it.Next()
+				if err != nil {
+					break
+				}
+
+				if instance.Labels != nil {
+					if _, hasTalis := instance.Labels["talis"]; hasTalis {
+						publicIP := ""
+						if len(instance.NetworkInterfaces) > 0 {
+							ni := instance.NetworkInterfaces[0]
+							if len(ni.AccessConfigs) > 0 && ni.AccessConfigs[0].NatIP != nil {
+								publicIP = *ni.AccessConfigs[0].NatIP
+							}
+						}
+
+						if cnt == 0 {
+							fmt.Printf("%-30s %-10s %-15s %-15s %s\n", "Name", "Status", "Zone", "Public IP", "Created")
+							fmt.Printf("%-30s %-10s %-15s %-15s %s\n", "----", "------", "------", "---------", "-------")
+						}
+
+						status := "UNKNOWN"
+						if instance.Status != nil {
+							status = *instance.Status
+						}
+						name := ""
+						if instance.Name != nil {
+							name = *instance.Name
+						}
+						created := ""
+						if instance.CreationTimestamp != nil {
+							created = *instance.CreationTimestamp
+						}
+
+						fmt.Printf("%-30s %-10s %-15s %-15s %s\n",
+							name,
+							status,
+							zone,
+							publicIP,
+							created)
+						cnt++
+					}
+				}
+			}
+		}
+	}
+
+	fmt.Println("Total number of talis instances:", cnt)
+	return nil
+}
+
+func (c *GCClient) GetConfig() Config {
+	return c.cfg
+}
+
+func NewGoogleCloudValidator(region string) Instance {
+	if region == "" || region == RandomRegion {
+		region = RandomGCRegion()
+	}
+	i := NewBaseInstance(Validator)
+	i.Provider = GoogleCloud
+	i.Slug = GCDefaultValidatorMachineType
+	i.Region = region
+	return i
+}
+
+func RandomGCRegion() string {
+	return GCRegions[rand.Intn(len(GCRegions))]
+}
+
+func gcClientOptions(cfg Config) ([]option.ClientOption, error) {
+	var opts []option.ClientOption
+	if cfg.GoogleCloudKeyJSONPath != "" {
+		keyJSON, err := os.ReadFile(cfg.GoogleCloudKeyJSONPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read Google Cloud key file at %s: %w", cfg.GoogleCloudKeyJSONPath, err)
+		}
+		opts = append(opts, option.WithCredentialsJSON(keyJSON))
+	}
+	return opts, nil
+}
+
+func RandomGCZone(region string) string {
+	zones, ok := GCZones[region]
+	if !ok || len(zones) == 0 {
+		return region + "-a"
+	}
+	return zones[rand.Intn(len(zones))]
+}
+
+func ensureGCFirewallRule(ctx context.Context, project string, opts []option.ClientOption) error {
```

Comment:
GC requires creating a firewall rule otherwise the servers will have all traffic blocked